synapse/handlers/room_member.py

```diff
@@ -235,9 +289,25 @@ async def _local_membership_update(
 
         if event.membership == Membership.JOIN:
             newly_joined = True
+            is_invite = False
             if prev_member_event_id:
                 prev_member_event = await self.store.get_event(prev_member_event_id)
                 newly_joined = prev_member_event.membership != Membership.JOIN
+                is_invite = prev_member_event.membership == Membership.INVITE
+
+            # If the member is not already in the room, check if they should be
+            # allowed access via membership in a space.
+            if (
+                newly_joined
+                and not is_invite
+                and not await self._can_join_restricted_room(
+                    prev_state_ids, event.room_version, user_id
+                )
+            ):
```

Comment:
This partially feels like we're re-implementing the auth checks here, but not sure there's a more reasonable way to do it.

Reply:
agreed, but :man_shrugging: